std/shared/src/main/scala/cats/effect/std/Hotswap2.scala

```diff
@@ -0,0 +1,166 @@
+
+object Hotswap2 {
+
+  /**
+   * Creates a new [[Hotswap2]] initialized with the specified resource, which represents a
+   * [[cats.effect.kernel.Resource]] that can be swapped during the lifetime of this
+   * [[Hotswap2]].
+   */
+  def apply[F[_], R](initial: Resource[F, R])(
+      implicit F: Concurrent[F]): Resource[F, Hotswap2[F, R]] =
+    Resource.eval(Semaphore[F](Long.MaxValue)).flatMap { semaphore =>
+      sealed abstract class State
+      case class Acquired(r: R, fin: F[Unit]) extends State
+      case object Finalized extends State
+
+      def initialize: F[Ref[F, State]] =
+        F.uncancelable { poll =>
+          poll(initial.allocated).flatMap {
+            case (r, fin) =>
+              exclusive.mapK(poll).onCancel(Resource.eval(fin)).surround {
+                F.ref(Acquired(r, fin))
+              }
+          }
+        }
+
+      def finalize(state: Ref[F, State]): F[Unit] =
+        state.getAndSet(Finalized).flatMap {
+          case Acquired(_, finalizer) => exclusive.surround(finalizer)
+          case Finalized => raise("NonEmptyHotswap already finalized")
+        }
+
+      def raise[A](message: String): F[A] =
+        F.raiseError[A](new IllegalStateException(message))
+
+      def exclusive: Resource[F, Unit] =
+        Resource.makeFull[F, Unit](poll => poll(semaphore.acquireN(Long.MaxValue)))(_ =>
+          semaphore.releaseN(Long.MaxValue))
+
+      Resource.make(initialize)(finalize).map { state =>
+        new Hotswap2[F, R] {
+
+          override def swap(next: Resource[F, R]): F[Unit] =
+            F.uncancelable { poll =>
+              poll(next.allocated).flatMap {
+                case (r, fin) =>
+                  exclusive.mapK(poll).onCancel(Resource.eval(fin)).surround {
+                    swapFinalizer(Acquired(r, fin))
+                  }
+              }
+            }
+
+          override def get: Resource[F, R] =
+            Resource.makeFull[F, R] { poll =>
+              poll(semaphore.acquire) *> // acquire shared lock
+                state.get.flatMap {
+                  case Acquired(r, _) => F.pure(r)
+                  case _ => raise("NonEmptyHotswap already finalized")
+                }
+            }(_ => semaphore.release)
+
+          private def swapFinalizer(next: State): F[Unit] =
+            state.flatModify {
+              case Acquired(_, fin) =>
+                next -> fin
+              case Finalized =>
+                val fin = next match {
+                  case Acquired(_, fin) => fin
+                  case _ => F.unit
+                }
+                Finalized -> (fin *> raise[Unit]("Cannot swap after finalization"))
+            }
+        }
+      }
+    }
+
+  /**
+   * Creates a `NonEmptyHotswap` that is largely functionally equivalent to `Hotswap`
+   *
+   * The primary difference is that [[Hotswap2.swap]] does not leak the resource.
+   */
+  def create[F[_], R](implicit F: Concurrent[F]): Resource[F, Hotswap2[F, Option[R]]] =
```

Comment:
Bikeshed? can also update the scaladoc.

```suggestion
  def empty[F[_], R](implicit F: Concurrent[F]): Resource[F, Hotswap2[F, Option[R]]] =
```